datashader/datatypes.py

```diff
@@ -0,0 +1,364 @@
+                if buffer_len <= max_supported:
+                    break
+
+            # infer dtype if not provided
+            if dtype is None:
+                dtype = np.result_type(*[np.atleast_1d(v)
+                                         for v in data
+                                         if v is not None])
+
+            # Initialize representation arrays
+            self._mask = np.zeros(index_len, dtype='bool')
+            self._start_indices = np.zeros(index_len, dtype=start_indices_dtype)
+            self._flat_array = np.zeros(buffer_len, dtype=dtype)
+
+            # Populate arrays
+            next_start_ind = 0
+            for i, array_el in enumerate(data):
+                # Check for null values
+                isnull = array_el is None
+
+                # Compute element length
+                n = len(array_el) if not isnull else 0
+
+                # Update mask
+                self._mask[i] = isnull
+
+                # Update start indices
+                self._start_indices[i] = next_start_ind
+
+                # Update flat array
+                self._flat_array[next_start_ind:next_start_ind+n] = array_el
+
+                # increment next start index
+                next_start_ind += n
+
+    @property
+    def flat_array(self):
+        """
+        numpy array containing concatenation of all nested arrays
+
+        Returns
+        -------
+        np.ndarray
+        """
+        return self._flat_array
+
+    @property
+    def mask(self):
+        """
+        boolean numpy array the same length as the ragged array where values
+        of True indicate missing values.
+
+        Returns
+        -------
+        np.ndarray
+        """
+        return self._mask
+
+    @property
+    def start_indices(self):
+        """
+        integer numpy array the same length as the ragged array where values
+        represent the index into flat_array where the corresponding ragged
+        array element begins.
+
+        Returns
+        -------
+        np.ndarray
+        """
+        return self._start_indices
+
+    def __len__(self):
+        """
+        Length of this array
+
+        Returns
+        -------
+        length : int
+        """
+        return len(self._start_indices)
+
+    def __getitem__(self, item):
+        """
+        Parameters
+        ----------
+        item : int, slice, or ndarray
+            * int: The position in 'self' to get.
+
+            * slice: A slice object, where 'start', 'stop', and 'step' are
+              integers or None
+
+            * ndarray: A 1-d boolean NumPy ndarray the same length as 'self'
+        """
+        if isinstance(item, Integral):
+            if item < -len(self) or item >= len(self):
+                raise IndexError(item)
+            elif self.mask[item]:
+                return None
+            else:
+                # Convert negative item index
+                if item < 0:
+                    item = 5 + item
```

Comment:
Sily question: why is this 5, and not `len(self)`?

Reply:
Silly answer: Copy paste error 😊

Reply:
Fixed in 5f46b8e85d6ddac4157ed4b047ed72a079b1aeab